PowerUp/app/src/main/java/powerup/systers/com/MapActivity.java

```diff
@@ -26,30 +30,37 @@
 
 public class MapActivity extends Activity {
 
-    private DatabaseHandler mDbHandler;
+    private DataSource dataSource;
     private OnClickListener onClickListener = new OnClickListener() {
         @Override
         public void onClick(View v) {
-            ImageView scenarioChooser = (ImageView) v;
+            final ImageView scenarioChooser = (ImageView) v;
             if (v.isEnabled()){
                 // checks whether has been already completed & then open GameActivity if setSessionId returns true
-                if (getmDbHandler().setSessionId(getScenarioName(scenarioChooser.getId()))) {
-                    startActivityForResult(new Intent(MapActivity.this, GameActivity.class), 0);
-                    overridePendingTransition(R.animator.fade_in_custom, R.animator.fade_out_custom);
-                } else if (new MinesweeperSessionManager(MapActivity.this).isMinesweeperOpened()) { //if minesweeper game was left incomplete
-                    startActivity(new Intent(MapActivity.this, MinesweeperGameActivity.class));
-                    overridePendingTransition(R.animator.fade_in_custom, R.animator.fade_out_custom);
-                } else if (new SinkToSwimSessionManager(MapActivity.this).isSinkToSwimOpened()) {
-                    startActivity(new Intent(MapActivity.this, SinkToSwimGame.class));
-                } else if (new VocabMatchSessionManager(MapActivity.this).isVocabMatchOpened()) {
-                    startActivity(new Intent(MapActivity.this, VocabMatchGameActivity.class));
-                } else {
-                    Intent intent = new Intent(MapActivity.this, ScenarioOverActivity.class);
-                    intent.putExtra(PowerUpUtils.SOURCE,PowerUpUtils.MAP);
-                    intent.putExtra(PowerUpUtils.SCENARIO_NAME,getScenarioName(scenarioChooser.getId()));
-                    startActivityForResult(intent, 0);
-                    overridePendingTransition(R.animator.fade_in_custom, R.animator.fade_out_custom);
-                }
+                dataSource.setSessionId(getScenarioName(scenarioChooser.getId()), new IDataSource.LoadBooleanCallback() {
```

Comment:
@manjumegh @manishakeim @niksj Need special review here. Is it good or need any changes?

Reply:
Its good. Big( and heavy) if-else-if ladder but that's okay. Its structured, so good. I don't feel any need of changes.